test/unittests/libp2p/net.cpp

```diff
@@ -939,6 +940,85 @@ BOOST_AUTO_TEST_CASE(findNodeIsSentAfterPong)
     BOOST_CHECK_EQUAL(datagram3->typeName(), "FindNode");
 }
 
+BOOST_AUTO_TEST_CASE(pingNotSentAfterPongForKnownNode)
+{
+    // Validate 3 scenarios:
+    // * A ping is sent after a pong for a new node
+    // * No ping is sent after a pong for a known node with a valid endpoint proof
+    // * A ping is sent after a pong for a known node with an invalid endpoint proof
+
+    TestNodeTableHost nodeTableHost1(0);
+    nodeTableHost1.start();
+    auto& nodeTable1 = nodeTableHost1.nodeTable;
+
+    TestUDPSocketHost nodeSocketHost2{getRandomPortNumber()};
+    nodeSocketHost2.start();
+    auto nodePort2 = nodeSocketHost2.port;
+    auto nodeEndpoint2 =
+        NodeIPEndpoint{bi::address::from_string(c_localhostIp), nodePort2, nodePort2};
+
+    // Add node to the node table to trigger ping
+    auto nodeKeyPair2 = KeyPair::create();
+    Node node2(nodeKeyPair2.pub(), nodeEndpoint2);
+    nodeTable1->addNode(node2);
+
+    // Verify ping is received
+    auto packetReceived1 = nodeSocketHost2.packetsReceived.pop(chrono::milliseconds(5000));
+    auto datagram1 =
+        DiscoveryDatagram::interpretUDP(bi::udp::endpoint{}, dev::ref(packetReceived1));
+    auto ping1 = dynamic_cast<PingNode const&>(*datagram1);
+    BOOST_CHECK_EQUAL(datagram1->typeName(), "Ping");
+
+    // Send pong to trigger endpoint proof in node table
+    Pong pong(nodeTable1->m_hostNodeEndpoint);
+    pong.echo = ping1.echo;
+    pong.sign(nodeKeyPair2.secret());
+    nodeSocketHost2.socket->send(pong);
+
+    // Receive FindNode packet
+    auto packetReceived3 = nodeSocketHost2.packetsReceived.pop(chrono::milliseconds(20000));
+    auto datagram3 =
+        DiscoveryDatagram::interpretUDP(bi::udp::endpoint{}, dev::ref(packetReceived3));
+    BOOST_CHECK_EQUAL(datagram3->typeName(), "FindNode");
+
+    // Ping the node table and verify that a pong is received but no ping is received after
+    // (since the endpoint proof has already been completed)
+    PingNode ping2(nodeEndpoint2, nodeTable1->m_hostNodeEndpoint);
+    ping2.sign(nodeKeyPair2.secret());
+    nodeSocketHost2.socket->send(ping2);
+
+    auto packetReceived4 = nodeSocketHost2.packetsReceived.pop(chrono::milliseconds(5000));
+    auto datagram4 =
+        DiscoveryDatagram::interpretUDP(bi::udp::endpoint{}, dev::ref(packetReceived4));
+    BOOST_CHECK_EQUAL(datagram4->typeName(), "Pong");
+
+    // Verify that the next packet received is not a ping
+    auto packetReceived5 = nodeSocketHost2.packetsReceived.pop(chrono::milliseconds(20000));
+    auto datagram5 =
+        DiscoveryDatagram::interpretUDP(bi::udp::endpoint{}, dev::ref(packetReceived5));
+    BOOST_REQUIRE(datagram5->typeName() != "Ping");
+
+    // Force the endpoint proof to be invalid
+    shared_ptr<NodeEntry> newNode = nodeTable1->m_allNodes.begin()->second;
```

Comment:
this is not thread-safe, can you use `nodeTable1->nodeEntry()` ?
(still would be not entirely safe as `lastPongReceivedTime` be accessed from another thread... but polluting the tests with perfect thread-safety seems too much of a hassle to me)